File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/content/classes/alchemist/choices/AlchemistFormulaeChoice.java
package ai.ragu.charactermanager.content.classes.alchemist.choices;

import ai.ragu.charactermanager.dto.CharacterDto;
import ai.ragu.charactermanager.dto.Choice;
import ai.ragu.charactermanager.dto.ProficiencyModifierDto;
import ai.ragu.charactermanager.enumeration.ProficiencyLevelEnum;
import ai.ragu.charactermanager.enumeration.SkillEnum;

import java.util.List;

public class AlchemistFormulaeChoice extends Choice {
    private int numAnswers;
    public AlchemistFormulaeChoice(String suffix, String description, List<String> options, int numAnswers) {
        this.code = AlchemistFormulaeChoice.class.getName() + suffix;
        this.numAnswers = numAnswers;
        this.description = description;
        this.options = options;
    }

    @Override
    public int getNumAnswers(CharacterDto characterDto) {
        return this.numAnswers;
    }

    @Override
    public List<Choice> callback(CharacterDto characterDto) {
        characterDto.clearSkillsModifiersGrantedBy(code);
        for (String answerStr : answers) {
            SkillEnum skillEnum = SkillEnum.fromValue(answerStr);
            ProficiencyModifierDto newModifier = new ProficiencyModifierDto(
                    code,
                    ProficiencyLevelEnum.TRAINED,
                    0
            );
            characterDto.getSkills().get(skillEnum).getProficiency().getModifiers().add(newModifier);
        }
        return List.of();
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/content/classes/alchemist/choices/AlchemistFormulaeChoice.kt
package ai.ragu.charactermanager.content.classes.alchemist.choices

import ai.ragu.charactermanager.dto.CharacterDto
import ai.ragu.charactermanager.dto.Choice
import ai.ragu.charactermanager.dto.ProficiencyModifierDto
import ai.ragu.charactermanager.enumeration.ProficiencyLevelEnum
import ai.ragu.charactermanager.enumeration.SkillEnum

class AlchemistFormulaeChoice(suffix: String, description: String?, options: List<String?>?, numAnswers: Int) : Choice() {
    private val numAnswers: Int
    override fun getNumAnswers(characterDto: CharacterDto): Int {
        return numAnswers
    }

    override fun callback(characterDto: CharacterDto): List<Choice?> {
        characterDto.clearSkillsModifiersGrantedBy(code!!)
        for (answerStr in answers!!) {
            val skillEnum: SkillEnum = SkillEnum.Companion.fromValue(answerStr)
            val newModifier = ProficiencyModifierDto(
                    code,
                    ProficiencyLevelEnum.TRAINED,
                    0
            )
            characterDto.skills[skillEnum].getProficiency().modifiers.add(newModifier)
        }
        return java.util.List.of()
    }

    init {
        code = AlchemistFormulaeChoice::class.java.name + suffix
        this.numAnswers = numAnswers
        this.description = description
        this.options = options
    }
}